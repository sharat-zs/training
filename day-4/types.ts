type StringOrNum = string | number

export let x: StringOrNum = 5
x = 'abc'

// @ts-expect-error
x = false

type Point2D = {
    x: number
    y: number
}

type PointAlias = {
    x: number,
    y: number
}

type Point3D = Point2D & {
    z: number
}

type Maybe3D = Point3D | Point2D

let p: Maybe3D = { x: 12, y: 12 }
let o: Maybe3D = { x: 12, y: 12, z: 100 }

let g: Point3D = { x: 12, y: 12, z: 30 }

let pointAlias: PointAlias = { x: 12, y: 12 }

interface IPointInterface {
    x: number
    y: number
}

interface IPoint3DInterface extends IPointInterface {
    z: number
}

let y: IPointInterface = {
    x: 10,
    y: 10
}

function getDistance(point: { x: number, y: number }) {
    return 10
}

function getDistanceFromCenter(point: Point2D | Point3D): number {
    if (is3D(point))
        return Math.sqrt(point.x * point.x + point.y * point.y + point.z * point.z)
    return Math.sqrt(point.x * point.x + point.y * point.y)
}

//type guard
function is3D(point: Point2D | Point3D): point is Point3D {
    return (point as Point3D).z !== undefined
}



