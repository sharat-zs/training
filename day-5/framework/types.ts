import { Server } from 'http'

export type Context = {
    headers?: Record<string, unknown>
    query?: Record<string, unknown>
    body?: Record<string, unknown>
    params?: Record<string, unknown>
    path: string
}

export type RouteHandler = (ctx: Context) => Promise<any>

export type RouteMethod = (path: string, handler: RouteHandler) => IApp

export interface IApp {
    get: RouteMethod
    put: RouteMethod
    post: RouteMethod
    delete: RouteMethod
    start: (port?: number) => Promise<Server>
}

export enum HttpMethod {
    Get = 'get',
    Post = 'post',
    Delete = 'delete',
    Put = 'put'
}

export enum HTTPStatus {
    OK = 200,
    CREATED = 201,
    ACCEPTED = 202,
    NO_CONTENT = 204,
    PARTIAL_CONTENT = 206,
    BAD_REQUEST = 400,
    UNAUTHORIZED = 401,
    FORBIDDEN = 403,
    NOT_FOUND = 404,
    METHOD_NOT_ALLOWED = 405,
    CONFLICT = 409,
    INTERNAL_SERVER_ERROR = 500,
    NOT_IMPLEMENTED = 501
}
