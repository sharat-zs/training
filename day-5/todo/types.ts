import { Context } from '../framework'

export enum TodoStatus {
    Pending = 'pending',
    InProgress = 'in-progress',
    Done = 'done'
}

export type TodoItem = {
    id: number
    title: string
    status: TodoStatus
}

export type CreateTodoBody = Omit<TodoItem, 'id'>

export interface ITodoController {
    getList: (ctx: Context) => Promise<TodoItem[]>
    createItem: (ctx: Context) => Promise<TodoItem>
}

export interface ITodoService {
    getList: () => Promise<TodoItem[]>
    createItem: (data: CreateTodoBody) => Promise<TodoItem>
}
