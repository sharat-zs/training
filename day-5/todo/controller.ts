import { Context, StandardError } from '../framework'
import { TodoService } from './service'
import { CreateTodoBody, ITodoController, ITodoService } from './types'

export class Controller implements ITodoController {
    static instance: ITodoController

    static getInstance() {
        if (this.instance === undefined)
            this.instance = new Controller(TodoService.getInstance())
        return this.instance
    }

    constructor(private readonly service: ITodoService) {}

    getList: (ctx: Context) => Promise<any> = async (_ctx) => {
        return this.service.getList()
    }

    createItem: (ctx: Context) => Promise<any> = async (ctx) => {
        const data = ctx.body
        if (data === undefined || !isValidCreateTodoBody(data))
            throw new StandardError('Bad request', 'BAD_REQUEST', 400)
        return this.service.createItem(data)
    }
}

function isValidCreateTodoBody(data: unknown): data is CreateTodoBody {
    return (
        (data as CreateTodoBody).title !== undefined &&
        (data as CreateTodoBody).status !== undefined
    )
}

// Achieving something similar using a more "mdoule" based approach
// const service = TodoService.getInstance()

// export const getList: (ctx: Context) => Promise<any> = async (_ctx) => {
//     return service.getList()
// }
