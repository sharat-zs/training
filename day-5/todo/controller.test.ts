import {
    createMockContext,
    MockTodoService,
    MOCK_TODO_ARRAY
} from './test-utils'
import { Controller } from './controller'
import { TodoStatus } from './types'

describe('TodoController', () => {
    const mockService = new MockTodoService()

    describe('getList', () => {
        it('should call service getList', async () => {
            mockService.getList.mockResolvedValueOnce(MOCK_TODO_ARRAY)
            const controller = new Controller(mockService)
            const res = await controller.getList(createMockContext())
            expect(mockService.getList).toHaveBeenCalledTimes(1)
            expect(res).toMatchSnapshot()
        })
    })

    describe('create item', () => {
        it('should reject invalid context', async () => {
            mockService.createItem.mockResolvedValueOnce(MOCK_TODO_ARRAY[0])
            const controller = new Controller(mockService)
            expect(
                controller.createItem(createMockContext())
            ).rejects.toMatchSnapshot()
            expect(mockService.createItem).not.toHaveBeenCalled()
        })

        it('should return created todoitem', async () => {
            mockService.createItem.mockResolvedValueOnce(MOCK_TODO_ARRAY[0])
            const controller = new Controller(mockService)
            const res = await controller.createItem(
                createMockContext({
                    body: { title: 'mock title', status: TodoStatus.Pending }
                })
            )
            expect(mockService.createItem).toHaveBeenCalledTimes(1)
            expect(res).toMatchSnapshot()
        })
    })

    describe('singleton test', () => {
        expect(Controller.getInstance()).toStrictEqual(Controller.getInstance())
    })
})
