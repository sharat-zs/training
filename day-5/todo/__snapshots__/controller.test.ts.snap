// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoController create item should reject invalid context 1`] = `[Error: Bad request]`;

exports[`TodoController create item should return created todoitem 1`] = `
Object {
  "id": 1,
  "status": "pending",
  "title": "task -1",
}
`;

exports[`TodoController getList should call service getList 1`] = `
Array [
  Object {
    "id": 1,
    "status": "pending",
    "title": "task -1",
  },
]
`;
